core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -800,6 +835,103 @@ private void updateStageExecutions()
+                RemoteSourceNode runtimeAdaptiveRemoteSourceNode = new RemoteSourceNode(
+                        getRuntimeAdaptivePlanNodeId(sourceRemoteSourceNode.getId()),
+                        sourceFragment.getId(),
+                        sourceFragment.getOutputPartitioningScheme().getOutputLayout(),
+                        sourceRemoteSourceNode.getOrderingScheme(),
+                        sourceRemoteSourceNode.getExchangeType(),
+                        sourceRemoteSourceNode.getRetryPolicy());
+                PlanFragment runtimeAdaptivePlanFragment = new PlanFragment(
+                        getRuntimeAdaptivePlanFragmentId(sourceFragment.getId()),
+                        runtimeAdaptiveRemoteSourceNode,
+                        sourceFragment.getSymbols(),
+                        FIXED_HASH_DISTRIBUTION,
+                        Optional.of(oldPartitionCount),
+                        ImmutableList.of(),
+                        sourceFragment.getOutputPartitioningScheme().withPartitionCount(Optional.of(newPartitionCount)),
+                        sourceFragment.getStatsAndCosts(),
+                        sourceFragment.getActiveCatalogs(),
+                        sourceFragment.getJsonRepresentation());
+                SubPlan newSource = new SubPlan(
+                        runtimeAdaptivePlanFragment,
+                        ImmutableList.of(source));
+                newSources.add(newSource);
+            }
+
+            return new SubPlan(
+                    planFragment
+                            .withPartitionCount(Optional.of(newPartitionCount))
+                            .withRoot(rewriteWith(new RemoteSourceNodeRuntimeAdaptiveRewriter(), planFragment.getRoot())),
+                    newSources.build());
+        }
+
+        private static class RemoteSourceNodeRuntimeAdaptiveRewriter
+            extends SimplePlanRewriter<Void>
+        {
+            @Override
+            public PlanNode visitRemoteSource(RemoteSourceNode node, RewriteContext<Void> context)
+            {
+                return node.withSourceFragmentIds(node.getSourceFragmentIds().stream().map(Scheduler::getRuntimeAdaptivePlanFragmentId).collect(toImmutableList()));
+            }
+        }
+
+        private static PlanNodeId getRuntimeAdaptivePlanNodeId(PlanNodeId planNodeId)
+        {
+            int planNodeIdAsInt = Integer.parseInt(planNodeId.toString());
+            checkArgument(planNodeIdAsInt < RUNTIME_ADAPTIVE_PLAN_NODE_ID_THRESHOLD,
+                    "planNodeIdAsInt %s larger than RUNTIME_ADAPTIVE_PLAN_NODE_ID_THRESHOLD %s",
+                    planNodeIdAsInt, RUNTIME_ADAPTIVE_PLAN_NODE_ID_THRESHOLD);
+            return new PlanNodeId(String.valueOf(RUNTIME_ADAPTIVE_PLAN_NODE_ID_THRESHOLD + planNodeIdAsInt));
+        }
+
+        private static PlanFragmentId getRuntimeAdaptivePlanFragmentId(PlanFragmentId fragmentId)
```

Comment:
Maybe just capture what is the biggest fragment id in plan so far and use sequance starting with that one.